packages/core/permissions/src/engine/index.ts

```diff
@@ -15,13 +15,16 @@ import { Permission } from '../domain/permission';
 
 export { abilities };
 
-type Provider = ReturnType<typeof providerFactory>;
+type Provider = Omit<ReturnType<typeof providerFactory>, 'register'> & {
+  register(...args: unknown[]): Promise<Provider> | Provider;
+};
+
 type ActionProvider = Provider;
 type ConditionProvider = Provider;
 
 export interface Engine {
   hooks: PermissionEngineHooks;
-  on(hook: HookName, handler: (...args: unknown[]) => unknown): Engine;
+  on(hook: HookName, handler: (...args: any[]) => any): Engine;
```

Comment:
This looks like a step in the wrong direction 😬 . Any chance a generic could work 🤔 ?

`handler: <T, R>(...args: T[]) => R): Engine`

Reply:
so you can't define it this way, but you can do 

```ts
on<TParameters extends any[], TResult>(
    hook: HookName,
    handler: (...args: TParameters) => TResult
  ): Engine;
```

so you can then do 

```ts
t.on<string[], string>('after-format::validate.permission', (permissions) => {
  return 'foo';
});
```

but this is useless unless we could define constratins on TParameters and TResult, which we can't because it will only be decided by the caller.

So the result is that you define the types of the function that you pass it which is itself defined by you anyway. In the end you are not enforcing anything 

```ts
t.on('after-format::validate.permission', (permissions: string) => {
  return 'foo';
});
```

Reply:
gotcha, alright thanks for taking the time to explore and explain :)